ZEP008-hot-backup.md

```diff
@@ -0,0 +1,619 @@
+
+Scenario 6:
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- checkpoint taken record is committed
+- Failover happens before BackupActor starts taking backup
+
+This is similar to scenario 4. The difference is that there is no "ongoing" backup. But the new leader will still mark it as failed because it will not take a new backup.
+
+Scenario 7:
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- checkpoint taken record is committed
+- New snapshot is taken with processedPosition > checkpoint position
+- Failover happens before BackupActor starts taking backup
+- New leader starts with the new snapshot at processedPosition > checkpoint position
+
+This is similar to scenario 6.
+
+Scenario 8:
+
+- StreamProcessor sends "takeBackup" command to BackupActor
+- StreamProcess writes "checkpoint taken" record
+- BackupActor starts taking snapshot
+- checkpoint taken record is committed
+- Backup fails due to some errors - such as i/o exceptions
+
+This is outside of our control. The backup can fail due to several reasons such as remote storage is not available, or other i/o errors. We can retry it for a few times before marking the backup as failed.
+
+##### Failure scenarios in coordinator to Zeebe broker communication
+
+- Backup request from coordinator to a Zeebe broker is lost.
+- Acknowledgment from Zeebe broker to the coordinator is lost.
+
+In this case, the coordinator can resend the backup request. It is ok if a partition receives the request twice. Only the first request triggers the backup. The second request will be processed by the streamProcessor, but it will not trigger a backup. Instead it can just send an acknowledgment back to the coordinator indicating that the backup is already ongoing.
+
+What happens if a backup fails?
+If a backup failed, the coordinator can see the status via the monitoring query. If a backup fails, there is no use in re-sending the backup request with the same backup id. The users must retry with a new backup id.
+
+##### Rationale for restore process
+The backup should contain the snapshot and the log until the checkpoint position. However, while taking the backup, raft will be writing new entries to the log concurrently. To simplify taking the backup, we include extra entries in the backup. Instead, during restore we will fix it by deleting all entries that should not be in the backup.
+
+However, after deleting the entries, according to StreamProcessor lastCheckpointId = backupId - 1. This is not correct, as it may trigger another backup with the same backupId. To prevent this, we write a new command restore backupId. This way, StreamProcessor on processing this command can just update the state and set lastcheckpointId = backupId.
```

Comment:
Even if it is manual, we have to provide a way to delete the extra records after the checkpoint position. If we want to prevent that, we have to take care of it while taking the backup. That is while taking the backup instead of simply copying the segments, it has to copy the records until the checkpoint position. This way during restore, we don't have to delete anything.